lib/services/user_login_state.go

```
@@ -0,0 +1,87 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package services
+
+import (
+	"context"
+
+	"github.com/gravitational/trace"
+
+	"github.com/gravitational/teleport/api/types/userloginstate"
+	"github.com/gravitational/teleport/lib/utils"
+)
+
+// UserLoginStateGetter defines an interface for reading user login states.
+type UserLoginStateGetter interface {
+	// GetUserLoginState returns the specified user login state resource.
+	GetUserLoginState(context.Context, string) (*userloginstate.UserLoginState, error)
+}
+
+// UserLoginStates defines an interface for managing user login states.
+type UserLoginStates interface {
+	UserLoginStateGetter
+
+	// UpsertUserLoginState creates or updates a user login state resource.
+	UpsertUserLoginState(context.Context, *userloginstate.UserLoginState) (*userloginstate.UserLoginState, error)
+
+	// DeletetUserLoginState deletes a user login state resource.
+	DeleteUserLoginState(context.Context, string) error
+}
+
+// MarshalUserLoginState marshals the user login state resource to JSON.
+func MarshalUserLoginState(userLoginState *userloginstate.UserLoginState, opts ...MarshalOption) ([]byte, error) {
+	if err := userLoginState.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	cfg, err := CollectOptions(opts)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if !cfg.PreserveResourceID {
+		copy := *userLoginState
+		copy.SetResourceID(0)
+		userLoginState = &copy
+	}
+	return utils.FastMarshal(userLoginState)
+}
+
+// UnmarshalUserLoginState unmarshals the user login state resource from JSON.
+func UnmarshalUserLoginState(data []byte, opts ...MarshalOption) (*userloginstate.UserLoginState, error) {
+	if len(data) == 0 {
+		return nil, trace.BadParameter("missing user login state data")
+	}
```

Ditch this and let the unmarshal fail? I'm not sure the special case is worth it.

```suggestion
```